File: Day11.java
import java.util.HashSet;
import java.util.ArrayList;

class Day11 {
  private final long startState =
      (1l | 128l | 256l | 512l | 32l | 64l | 4096l | 8192l)
    | ((2l | 4l) << floorShifts[1])
    | ((8l | 16l | 1024l | 2048l) << floorShifts[2]);

  public static final long maxItem = 8192;
  public static final long allMcs = 1 | 2 | 4 | 8 | 16 | 32 | 64;
  public static final long allGens = 128 | 256 | 512 | 1024 | 2048 | 4096 | 8192;
  public static final long finalSolution = allMcs | allGens;

  public static final long[] floorShifts = new long[] {0, 14, 28, 42};
  public static final long elevatorShift = 56;
  public static final long microchipShift = 0;
  public static final long generatorShift = 7;

  public static final long elevatorMask = 0x03l;
  public static final long allFloorsMask = 0x00FF_FFFF_FFFF_FFFFl;
  public static final long oneFloorMask = 0x3FFFl;
  public static final long halfFloorMask = 0x7Fl;

  private final HashSet<Long> seenStates;
  private ArrayList<Runner> currentRunners;

  private class Runner {
    private final long currentState;

    public Runner(final long state) {
      this.currentState = state;
    }

    public boolean makeRunners(ArrayList<Runner> nextRunners) {
      // printState(this.currentState);

      final int floorIndex = Day11.getElevator(this.currentState);
      final long floorItems = Day11.getFloorItems(this.currentState, floorIndex);

      if (floorIndex == 3 && floorItems == Day11.finalSolution) {
        return true;
      }

      final int upFloor = floorIndex + 1;
      final int downFloor = floorIndex - 1;

      final boolean canGoUp = floorIndex < 3;
      final boolean canGoDown = floorIndex > 0;

      for (long i = 1; i <= maxItem; i = i << 1) {
        if ((floorItems & i) == 0) {
          continue;
        }

        // start at i, because i can travel alone
        // when j == i, it is basically i travelling alone. helps reduce code
        for (long j = i; j <= maxItem; j = j << 1) {
          if ((floorItems & j) == 0) {
            continue;
          }

          final long ij = i | j;

          if (canGoUp) {
            final long nextState = Day11.makeNewState(this.currentState, ij, floorIndex, upFloor);

            if (wouldStateBeOk(nextState) && addVisitedStateIfUnvisited(nextState)) {
              nextRunners.add(new Runner(nextState));
            }
          }
          if (canGoDown) {
            final long nextState = Day11.makeNewState(this.currentState, ij, floorIndex, downFloor);

            if (wouldStateBeOk(nextState) && addVisitedStateIfUnvisited(nextState)) {
              nextRunners.add(new Runner(nextState));
            }
          }
        }
      }

      return false;
    }
  }

  public Day11() {
    seenStates = new HashSet<Long>();
    currentRunners = new ArrayList<Runner>();
    currentRunners.add(new Runner(startState));
  }

  public int run() {
    int iterations = 0;

    while (this.currentRunners.size() > 0) {
      ArrayList<Runner> nextRunners = new ArrayList<Runner>();
      for (Runner runner : this.currentRunners) {
        if (runner.makeRunners(nextRunners)) {
          return iterations;
        }
      }
      iterations++;
      System.out.printf("iter: %d %d\n", iterations, nextRunners.size());
      this.currentRunners = nextRunners;
      // break;
    }

    return -1;
  }

  public boolean addVisitedStateIfUnvisited(final long state) {
    // printState(state);
    return this.seenStates.add(state);
  }

  public static void main(String[] args) {
    final Day11 solver = new Day11();
    final int part2Solution = solver.run();
    System.out.printf("Part 2 solution: %d\n", part2Solution);
  }

  // utility functions
  public static void printState(final long state) {
    System.out.printf("Elevator at: %d\n", getElevator(state) + 1);
    for (int i = 0; i < 4; i++) {
      System.out.printf("Floor %d:\n", i+1);
      printFloorState(getFloorItems(state, i));
    }
  }

  public static void printFloorState(final long floorItems) {
    if ((floorItems & 1) != 0) {
      System.out.println("thulium microchip");
    }
    if ((floorItems & 2) != 0) {
      System.out.println("plutonium microchip");
    }
    if ((floorItems & 4) != 0) {
      System.out.println("strontium microchip");
    }
    if ((floorItems & 8) != 0) {
      System.out.println("promethium microchip");
    }
    if ((floorItems & 16) != 0) {
      System.out.println("ruthenium microchip");
    }
    if ((floorItems & 32) != 0) {
      System.out.println("elerium microchip");
    }
    if ((floorItems & 64) != 0) {
      System.out.println("dilithium microchip");
    }

    if ((floorItems & 128) != 0) {
      System.out.println("thulium generator");
    }
    if ((floorItems & 256) != 0) {
      System.out.println("plutonium generator");
    }
    if ((floorItems & 512) != 0) {
      System.out.println("strontium generator");
    }
    if ((floorItems & 1024) != 0) {
      System.out.println("promethium generator");
    }
    if ((floorItems & 2048) != 0) {
      System.out.println("ruthenium generator");
    }
    if ((floorItems & 4096) != 0) {
      System.out.println("elerium generator");
    }
    if ((floorItems & 8192) != 0) {
      System.out.println("dilithium generator");
    }
  }

  public static long getFloorItems(final long state, final int floorIndex) {
    final long shift = floorShifts[floorIndex];
    return (state >> shift) & oneFloorMask;
  }

  public static int getElevator(final long state) {
    return (int)((state >> elevatorShift) & elevatorMask);
  }

  public static long makeNewState(final long oldState, final long changedItems, final int oldFloor, final int newFloor) {
    final long oldFloorShift = floorShifts[oldFloor];
    final long newFloorShift = floorShifts[newFloor];

    final long elevatorState = (long)newFloor << elevatorShift;

    final long xorBy = (changedItems << oldFloorShift) | (changedItems << newFloorShift);
    final long newStateNoElevator = (oldState ^ xorBy) & allFloorsMask;

    return newStateNoElevator | elevatorState;
  }

  // public static boolean hasMatchingGenerator(final long groupItems, final long mcId) {
  //   final long genId = mcId << generatorShift;
  //   return (groupItems & genId) != 0;
  // }
  // public static boolean hasMatchingMicrochip(final long groupItems, final long genId) {
  //   final long mcId = genId >> generatorShift;
  //   return (groupItems & mcId) != 0;
  // }

  public static boolean wouldGroupBeOk(final long groupItems) {
    final long gens = (groupItems >> generatorShift) & halfFloorMask;

    if (gens == 0) {
      return true;
    }

    final long mcs = groupItems & halfFloorMask;

    // when gen is 1, then we ignore that mc
    // when gen is 0, we need to know if the mc exists. basically AND mc with NOT gen
    return (mcs & ~(gens)) == 0;
  }

  public static boolean wouldStateBeOk(final long state) {
    for (int i = 0; i < 4; i++) {
      final long floorItems = getFloorItems(state, i);
      if (!wouldGroupBeOk(floorItems)) {
        return false;
      }
    }

    return true;
  }

}

File: day11.js
/*
The first floor contains a thulium generator, a thulium-compatible microchip, a plutonium generator, and a strontium generator.
The second floor contains a plutonium-compatible microchip and a strontium-compatible microchip.
The third floor contains a promethium generator, a promethium-compatible microchip, a ruthenium generator, and a ruthenium-compatible microchip.
The fourth floor contains nothing relevant.
*/

// 1 = typeT MC thulium
// 2 = typeP MC plutonium
// 4 = typeS MC strontium
// 8 = typeM MC promethium
// 16 = typeR MC ruthenium
// 32 = typeE MC elerium
// 64 = typeD MC dilithium

// 256 = typeT G thulium
// 512 = typeP G plutonium
// 1024 = typeS G strontium
// 2048 = typeM G promethium
// 4096 = typeR G ruthenium
// 8192 = typeE G elerium
// 16384 = typeD G dilithium

// if we had 64-bit... we could have done this. but nooo, JS is 32-bit.
// const startfloors = (1 | 256 | 512 | 1024 | 32 | 64 | 8192 | 16384) | ((2 | 4) << 16) | ((8 | 16 | 2048 | 4096) << 32);
const startfloors = [
  (1 | 256 | 512 | 1024 | 32 | 64 | 8192 | 16384),
  (2 | 4),
  (8 | 16 | 2048 | 4096),
  0
];

const allMcs = 1 | 2 | 4 | 8 | 16 | 32 | 64;
const allGens = 256 | 512 | 1024 | 2048 | 4096 | 8192 | 16384;
const finalSolution = allMcs | allGens;

let runners = [{current: 0, floors: startfloors, prev: []}];

// using an array of hashmaps for floors. odd, but eh. itll work. first level is elevator, second level is f0, ...
let prevStates = [{}, {}, {}, {}];

function getfloor(floors, floor) {
  // const shift = floor * 16;
  // return (floors >> shift) & 0xFFFF;
  return floors[floor];
}

function makeFloors(prev, newItems, oldFloor, newFloor) {
  // const xorBy = (newItems << (floor * 16)) | (newItems << (oldFloor * 16));
  // return prev ^ xorBy;

  let newFloors = prev.slice();

  if ((newFloors[newFloor] & newItems) !== 0) {
    throw new Error('this shouldnt happen');
  }

  newFloors[oldFloor] = newFloors[oldFloor] ^ newItems;
  newFloors[newFloor] = newFloors[newFloor] | newItems;

  return newFloors;
}

function addPath(next, floor) {
  // to reduce the number of hashmaps we have, going to use the full 32-bit int space for the hash
  // we use 2 hashes, so it is a hash in a hash in an array
  const hash1 = next[0] | (next[1] << 8);
  const hash2 = next[2] | (next[3] << 8);

  let pv = prevStates[floor];
  let v = pv;

  v = pv[hash1];
  if (!v) {
    pv[hash1] = {};
    v = pv[hash1];
  }
  pv = v;

  if (!pv[hash2]) {
    pv[hash2] = true;
  }
}

function isDoneBefore(next, floor) {
  const hash1 = next[0] | (next[1] << 8);
  const hash2 = next[2] | (next[3] << 8);

  const v = prevStates[floor][hash1];
  if (!v) {
    return false;
  }

  return v[hash2] === true;
}

function hasMatchingGenerator(items, mc) {
  const gen = mc << 8;
  return (items & gen) !== 0;
}
function hasMatchingMicrochip(items, gen) {
  const mc = gen >> 8;
  return (items & mc) !== 0;
}

function hasAnyGenerator(items) {
  return (items & allGens) !== 0;
}

function wouldGroupBeOk(items) {
  // if one generator exists:
  // then every microchip must have a matching generator

  if (!hasAnyGenerator(items)) {
    return true;
  }

  if ((items & 1) !== 0 && !hasMatchingGenerator(items, 1)) {
    return false;
  }
  if ((items & 2) !== 0 && !hasMatchingGenerator(items, 2)) {
    return false;
  }
  if ((items & 4) !== 0 && !hasMatchingGenerator(items, 4)) {
    return false;
  }
  if ((items & 8) !== 0 && !hasMatchingGenerator(items, 8)) {
    return false;
  }
  if ((items & 16) !== 0 && !hasMatchingGenerator(items, 16)) {
    return false;
  }
  if ((items & 32) !== 0 && !hasMatchingGenerator(items, 32)) {
    return false;
  }
  if ((items & 64) !== 0 && !hasMatchingGenerator(items, 64)) {
    return false;
  }

  return true;
}

function wouldAllFloorsBeOk(floors) {
  return floors.every(wouldGroupBeOk);
}

function canTravelTogether(items) {
  return wouldGroupBeOk(items);
}

function run(nextrunners, runner) {
  const floor = runner.current;
  const floorItems = getfloor(runner.floors, floor);

  if (floor === 3 && floorItems === finalSolution) {
    console.log('part 1 done');
    process.exit(0);
    return;
  }

  const upFloor = floor + 1;
  const downFloor = floor - 1;

  const canGoUp = floor < 3;
  const canGoDown = floor > 0;

  let addedOne = false;

  for (let i = 1; i <= 16384; i = i << 1) {
    if ((floorItems & i) === 0) {
      continue;
    }

    // start at i, because i can travel alone
    // when j === i, it is basically i travelling alone. helps reduce code
    for (let j = i; j <= 16384; j = j << 1) {
      if ((floorItems & j) === 0) {
        continue;
      }

      const ij = i | j;

      if (canGoUp) {
        const nextFloors = makeFloors(runner.floors, ij, floor, upFloor);
        if (wouldAllFloorsBeOk(nextFloors) && !isDoneBefore(nextFloors, upFloor)) {
          addedOne = true;
          nextrunners.push({
            current: upFloor,
            floors: nextFloors
          });
          addPath(nextFloors, upFloor);
        }
      }
      if (canGoDown) {
        const nextFloors = makeFloors(runner.floors, ij, floor, downFloor);
        if (wouldAllFloorsBeOk(nextFloors) && !isDoneBefore(nextFloors, downFloor)) {
          addedOne = true;
          nextrunners.push({
            current: downFloor,
            floors: nextFloors
          });
          addPath(nextFloors, downFloor);
        }
      }
    }
  }

  return nextrunners;
}

let iterations = 0;

while (runners.length) {
  runners = runners.reduce(run, []);
  console.log('iter:', ++iterations, runners.length);
}

console.log('no solution');
